src/wallet/test/walletdb_tests.cpp

```diff
@@ -48,10 +48,16 @@ BOOST_AUTO_TEST_CASE(walletdb_read_write_deadlock)
         LOCK(wallet->cs_wallet);
         auto legacy_spkm = wallet->GetOrCreateLegacyScriptPubKeyMan();
         BOOST_CHECK(legacy_spkm->SetupGeneration(true));
-        wallet->Flush();
+        {
+            // Keep the batch alive only for this call
+            const auto& db_batch = wallet->GetDatabase().MakeBatch();
+            BOOST_CHECK(HasLegacyRecords(wallet.get(), *db_batch));
+            wallet->Flush();
 
-        // Now delete all records, which performs a read write operation.
-        BOOST_CHECK(wallet->GetLegacyScriptPubKeyMan()->DeleteRecords());
+            // Now delete all records, which performs a read write operation.
+            BOOST_CHECK(wallet->GetLegacyScriptPubKeyMan()->DeleteRecords());
+            BOOST_CHECK(!HasLegacyRecords(wallet.get(), *db_batch));
+        }
```

Comment:
You didn't change the behavior. The `db_batch` object is alive **while** `Flush()` and `DeleteRecords()` are called. My suggestion was to scope each single `HasLegacyRecords()` call so it does not interfere with them.

Reply:
Yup, fixed it, thanks!